### PR TITLE
[stable/tomcat] allow more customizations

### PR DESCRIPTION
**Pull request by hlwanghl**
#### What this PR does / why we need it:

1) allow users inject more configurations, e.g. Tomcat global web.xml

2) allow users specify environment variables, which is useful when deploying same chart in different environments, and web apps can behave accordingly based-on different configurations (e.g. [Spring Boot profiles](https://docs.spring.io/spring-boot/docs/current/reference/html/howto-properties-and-configuration.html#howto-set-active-spring-profiles) can be chosen by env vars)

3) the previous hard-coded `hostPort` will prevent it to deploy more than one tomcat chart on one node; with this PR, it's removed and users can deploy as many tomcat charts on one node as he/she wants.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

File: stable/tomcat/Chart.yaml [hunk 1]
```diff
@@ -2,8 +2,8 @@ apiVersion: v1
 name: tomcat
 description: Deploy a basic tomcat application server with sidecar as web archive container
 icon: http://tomcat.apache.org/res/images/tomcat.png
-version: 0.2.0
-appVersion: 7.0
+version: 0.3.0
+appVersion: "7.0"
```
Comment (paulczar) [inline stable/tomcat/Chart.yaml:6]: why are you quoting the appVersion ?

File: stable/tomcat/templates/appsrv.yaml [hunk 2]
```diff
@@ -33,16 +36,29 @@ spec:
...
             - name: app-volume
               mountPath: {{ .Values.deploy.directory }}
+            {{- if .Values.extraVolumeMounts }}
+{{ tpl .Values.extraVolumeMounts . | indent 12 }}
+            {{- end }}
           ports:
             - containerPort: {{ .Values.service.internalPort }}
-              hostPort: 8009
+              {{- with .Values.hostPort }}
+              hostPort: {{ . }}
```
Comment (paulczar) [inline stable/tomcat/templates/appsrv.yaml:60]: I believe you're modifying the default behavior of the chart here by removing the default of `hostPort: 8009`.  Either leave the default as is, or you'll need to bump this as a breaking change.

File: stable/tomcat/values.yaml [hunk 3]
```diff
@@ -35,6 +37,24 @@ ingress:
   #    hosts:
   #      - chart-example.local
 
+env: |
```
Comment (paulczar) [inline stable/tomcat/values.yaml]: You shouldn't need these as embedded documents.  You can do the following to demonstrate intent in values file which is how this kind of thing is usually handled ( see the `annotations: {}` line in ingress )

```
env: []
  # - name: env
  #   value: test
```

**Review by paulczar — CHANGES_REQUESTED** [review]

File: stable/tomcat/Chart.yaml (hunk 1 above)
Comment (hlwanghl) [reply stable/tomcat/Chart.yaml:6]: If you take a look at the generated [index.yaml file](https://kubernetes-charts.storage.googleapis.com/index.yaml), the unquoted `7.0` (a decimal number) will be cut to `"7"` unexpectedly.

image

File: stable/tomcat/templates/appsrv.yaml (hunk 2 above)
Comment (hlwanghl) [reply stable/tomcat/templates/appsrv.yaml:60]: What about put it in values.yaml as a default value? This way, I can override it with empty value in custom values.

values.yaml:
```
hostPort: 8009
```

custom-values.yaml:
```
hostPort:
```

File: stable/tomcat/values.yaml (hunk 3 above)
Comment (hlwanghl) [reply stable/tomcat/values.yaml]: OK. I'll update accordingly.

**Comment by hlwanghl** [comment]
@paulczar addressed comments accordingly, and could you take a look? Thanks!

File: stable/tomcat/values.yaml [hunk 4]
```diff
@@ -35,6 +37,24 @@ ingress:
   #    hosts:
   #      - chart-example.local
 
+env: []
+  # - name: env
+  #   value: test
+
+extraVolumes: |
```
Comment (paulczar) [inline stable/tomcat/values.yaml]: this should be an empty array like `env: []` above it.

File: stable/tomcat/values.yaml [hunk 5]
```diff
@@ -35,6 +37,24 @@ ingress:
...
+  # - name: extra
+  #   emptyDir: {}
+
+extraVolumeMounts: |
```
Comment (paulczar) [inline stable/tomcat/values.yaml]: this should be an empty array like `env: []` above it.

File: stable/tomcat/values.yaml [hunk 6]
```diff
@@ -35,6 +37,24 @@ ingress:
...
+  # - name: extra
+  #   mountPath: /usr/local/tomcat/webapps/app
+  #   readOnly: true
+
+extraInitContainers: |
```
Comment (paulczar) [inline stable/tomcat/values.yaml]: this should be an empty array like `env: []` above it.

**Review by paulczar — CHANGES_REQUESTED** [review]

File: stable/tomcat/values.yaml (hunk 6 above)
Comment (hlwanghl) [reply stable/tomcat/values.yaml]: Thanks. Addressed.

File: stable/tomcat/values.yaml (hunk 5 above)
Comment (hlwanghl) [reply stable/tomcat/values.yaml]: Thanks. Addressed.

File: stable/tomcat/values.yaml (hunk 4 above)
Comment (hlwanghl) [reply stable/tomcat/values.yaml]: Thanks. Addressed.

**Pull request merged**